// @shopify/cli/dist/chunk-UNPXLODI.js
import {
  basename,
  dirname,
  extname,
  isAbsolute,
  join,
  normalize,
  relative,
  resolve
} from "./chunk-T5LLJYYS.js";
import {
  __commonJS,
  __require,
  __toESM,
  init_cjs_shims
} from "./chunk-ZPL24Y2D.js";

// ../../node_modules/.pnpm/commondir@1.0.1/node_modules/commondir/index.js
var require_commondir = __commonJS({
  "../../node_modules/.pnpm/commondir@1.0.1/node_modules/commondir/index.js"(exports, module) {
    init_cjs_shims();
    var path = __require("path");
    module.exports = function(basedir, relfiles) {
      if (relfiles) {
        var files = relfiles.map(function(r) {
          return path.resolve(basedir, r);
        });
      } else {
        var files = basedir;
      }
      var res = files.slice(1).reduce(function(ps, file) {
        if (!file.match(/^([A-Za-z]:)?\/|\\/)) {
          throw new Error("relative path without a basedir");
        }
        var xs = file.split(/\/+|\\+/);
        for (var i = 0; ps[i] === xs[i] && i < Math.min(ps.length, xs.length); i++)
          ;
        return ps.slice(0, i);
      }, files[0].split(/\/+|\\+/));
      return res.length > 1 ? res.join("/") : "/";
    };
  }
});

// ../cli-kit/src/public/node/path.ts
init_cjs_shims();
var import_commondir = __toESM(require_commondir(), 1);
import { fileURLToPath } from "url";
function joinPath(...paths) {
  return join(...paths);
}
function normalizePath(path) {
  return normalize(path);
}
function resolvePath(...paths) {
  return resolve(...paths);
}
function relativePath(from, to) {
  return relative(from, to);
}
function isAbsolutePath(path) {
  return isAbsolute(path);
}
function dirname2(path) {
  return dirname(path);
}
function basename2(path, ext) {
  return basename(path, ext);
}
function extname2(path) {
  return extname(path);
}
function relativizePath(path, dir = cwd()) {
  const result = (0, import_commondir.default)([path, dir]);
  const relativePath2 = relative(dir, path);
  const relativeComponents = relativePath2.split("/").filter((component) => component === "..").length;
  if (result === "/" || relativePath2 === "" || relativeComponents > 2) {
    return path;
  } else {
    return relativePath2;
  }
}
function moduleDirectory(moduleURL) {
  return dirname2(fileURLToPath(moduleURL));
}
function cwd() {
  return normalize(process.env.INIT_CWD ? process.env.INIT_CWD : process.cwd());
}
function sniffForPath(argv = process.argv) {
  const pathFlagIndex = argv.indexOf("--path");
  if (pathFlagIndex === -1) {
    const pathArg = argv.find((arg) => arg.startsWith("--path="));
    return pathArg?.split("=")[1];
  }
  const pathFlag = argv[pathFlagIndex + 1];
  if (!pathFlag || pathFlag.startsWith("-"))
    return;
  return pathFlag;
}
function sniffForJson(argv = process.argv) {
  const jsonFlagIndex = argv.indexOf("--json");
  return jsonFlagIndex !== -1;
}

export {
  require_commondir,
  joinPath,
  normalizePath,
  resolvePath,
  relativePath,
  isAbsolutePath,
  dirname2 as dirname,
  basename2 as basename,
  extname2 as extname,
  relativizePath,
  moduleDirectory,
  cwd,
  sniffForPath,
  sniffForJson
};
