// @shopify/cli/dist/node-5XXX7TNJ.js
import {
  findPort
} from "./chunk-Q4XGXPUG.js";
import {
  DEFAULT_INSPECTOR_PORT,
  H2O_BINDING_NAME,
  SUBREQUEST_PROFILER_ENDPOINT,
  createLogRequestEvent,
  getUtilityBannerlines,
  handleDebugNetworkRequest,
  handleMiniOxygenImportFail,
  importLocal,
  logRequestLine,
  setConstructors
} from "./chunk-HJ4AUFE7.js";
import {
  AbortError,
  outputNewline,
  readFile,
  renderSuccess
} from "./chunk-PIB2FEJJ.js";
import "./chunk-EQPYUHNM.js";
import "./chunk-6UDFXWNE.js";
import "./chunk-EARPFFS7.js";
import "./chunk-S4VBXFXP.js";
import "./chunk-LC6MEJT3.js";
import "./chunk-DL3MKXVR.js";
import "./chunk-FMKVOTQK.js";
import "./chunk-RUSG4O2Z.js";
import "./chunk-PRJ3U7MR.js";
import {
  source_default
} from "./chunk-WAEWSGBK.js";
import "./chunk-JIJ65JT6.js";
import "./chunk-WADS2TV5.js";
import "./chunk-GEN4RXTD.js";
import "./chunk-BYC3NA2Y.js";
import "./chunk-SAR5HGRG.js";
import "./chunk-NABCTJNW.js";
import "./chunk-YC4FLMDD.js";
import "./chunk-UNPXLODI.js";
import "./chunk-LSCXQWNO.js";
import "./chunk-K6257PKA.js";
import "./chunk-XQFDS44U.js";
import "./chunk-AMAYT2P2.js";
import "./chunk-VM4POVNV.js";
import "./chunk-ZPFJOUKZ.js";
import "./chunk-T5LLJYYS.js";
import {
  init_cjs_shims
} from "./chunk-ZPL24Y2D.js";

// ../../node_modules/.pnpm/@shopify+cli-hydrogen@0.0.0-next-9eb60d7-20240607102913_react-dom@17.0.2_react@17.0.2/node_modules/@shopify/cli-hydrogen/dist/lib/mini-oxygen/node.js
init_cjs_shims();
import { AsyncLocalStorage } from "node:async_hooks";
async function startNodeServer({
  appPort,
  watch = false,
  buildPathWorkerFile,
  buildPathClient,
  env,
  debug = false,
  inspectorPort,
  root
}) {
  const { startServer, Request, Response } = await importLocal(
    "@shopify/mini-oxygen/node",
    root
  ).catch(handleMiniOxygenImportFail);
  setConstructors({ Response });
  const logRequestEvent = createLogRequestEvent();
  const asyncLocalStorage = new AsyncLocalStorage();
  const serviceBindings = {
    [H2O_BINDING_NAME]: {
      fetch: async (request) => logRequestEvent(
        new Request(request.url, {
          method: "POST",
          body: JSON.stringify({
            ...asyncLocalStorage.getStore(),
            ...await request.json()
          })
        })
      )
    }
  };
  if (debug) {
    if (!inspectorPort)
      inspectorPort = await findPort(DEFAULT_INSPECTOR_PORT);
    (await import("node:inspector")).open(inspectorPort);
  }
  const readWorkerFile = () => readFile(buildPathWorkerFile).catch((error) => {
    throw new AbortError(
      `Could not read worker file.

` + error.stack,
      "Did you build the project?"
    );
  });
  const miniOxygen = await startServer({
    script: await readWorkerFile(),
    workerFile: buildPathWorkerFile,
    assetsDir: buildPathClient,
    publicPath: "",
    port: appPort,
    watch,
    autoReload: watch,
    modules: true,
    env: {
      ...env,
      ...process.env,
      ...serviceBindings
    },
    log: () => {
    },
    async onRequest(request, defaultDispatcher) {
      const url = new URL(request.url);
      if (url.pathname === SUBREQUEST_PROFILER_ENDPOINT) {
        return handleDebugNetworkRequest(request);
      }
      const requestId = request.headers.get("request-id");
      const startTimeMs = Date.now();
      const response = await asyncLocalStorage.run(
        { requestId, purpose: request.headers.get("purpose") },
        () => defaultDispatcher(request)
      );
      const endTimeMs = Date.now();
      logRequestLine({
        request: {
          url: request.url,
          method: request.method,
          headers: Object.fromEntries(request.headers.entries())
        },
        response: {
          status: response.status,
          statusText: response.statusText,
          headers: Object.fromEntries(response.headers.entries())
        },
        meta: {
          startTimeMs,
          endTimeMs,
          durationMs: startTimeMs > 0 ? endTimeMs - startTimeMs : 0
        }
      });
      return response;
    }
  });
  const listeningAt = `http://localhost:${miniOxygen.port}`;
  return {
    listeningAt,
    port: miniOxygen.port,
    async reload(options) {
      const nextOptions = {};
      if (options?.env) {
        nextOptions.env = {
          ...options.env,
          ...process.env
        };
      }
      nextOptions.script = await readWorkerFile();
      await miniOxygen.reload(nextOptions);
    },
    showBanner(options) {
      outputNewline();
      const customSections = [];
      if (options?.host) {
        customSections.push({ body: getUtilityBannerlines(options.host) });
      }
      if (debug && inspectorPort) {
        customSections.push({
          body: { warn: `Debugger listening on ws://localhost:${inspectorPort}` }
        });
      }
      renderSuccess({
        headline: `${options?.headlinePrefix ?? ""}MiniOxygen (Node Sandbox) ${options?.mode ?? "development"} server running.`,
        body: [
          `View ${options?.appName ? source_default.cyan(options?.appName) : "Hydrogen"} app:`,
          { link: { url: options?.host || listeningAt } }
        ],
        customSections
      });
      console.log("");
    },
    async close() {
      await miniOxygen.close();
    }
  };
}
export {
  startNodeServer
};
